### PR TITLE
Add HeapDumpOnOutOfMemoryError for worker daemons

### PR DESCRIPTION
**Pull request by blindpirate**
This closes https://github.com/gradle/gradle-private/issues/2950

We add HeapDumpOnOutOfMemoryError for all worker daemons.

### CODE REVIEW

File: buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt [hunk 1]
```diff
@@ -156,6 +157,7 @@ class UnitTestAndCompilePlugin : Plugin<Project> {
         options.isFork = true
         options.encoding = "utf-8"
         options.isIncremental = true
+        options.forkOptions.jvmArgs?.add("-XX:+HeapDumpOnOutOfMemoryError")
```
Comment (wolfs) [inline buildSrc/subprojects/plugins/src/main/kotlin/org/gradle/gradlebuild/unittestandcompile/UnitTestAndCompilePlugin.kt]: The `options.forkOptions` should be merged with the `groovyOptions.forkOptions`, so no need to add it twice.
https://github.com/gradle/gradle/blob/02aabe437352a8625f07488f8553ea695e9d5ee4/subprojects/language-groovy/src/main/java/org/gradle/api/internal/tasks/compile/daemon/DaemonGroovyCompiler.java#L96

File: buildSrc/build.gradle.kts [hunk 2]
```diff
@@ -207,6 +207,7 @@ fun Project.applyGroovyProjectConventions() {
     tasks.withType<GroovyCompile>().configureEach {
         groovyOptions.apply {
             encoding = "utf-8"
+            forkOptions.jvmArgs?.add("-XX:+HeapDumpOnOutOfMemoryError")
```
Comment (wolfs) [inline buildSrc/build.gradle.kts]: I suppose we don't need the heap dump in buildSrc, right? We can leave the arguments there, though. Any idea where the heap dump will end up?

Comment (blindpirate) [reply buildSrc/build.gradle.kts]: Heap dump of worker daemo will be in `~/.gradle/worker` and heap dump of buildSrc (Gradle daemon) will be in `~/.gradle/daemon`.

**Comment by blindpirate** [comment]
Thanks @wolfs !

**Pull request merged**